tests/unit/optimization/SGDMinimizer_unittest.cc

```diff
@@ -0,0 +1,458 @@
+
+	x(0,0)=3.18934210549; x(0,1)=-6.36734839959; x(0,2)=3.87646568343;
+	x(1,0)=9.71965286623; x(1,1)=-6.49372199537; x(1,2)=7.85930917808;
+	x(2,0)=0.590611116182; x(2,1)=-9.78182994856; x(2,2)=8.12538297323;
+	x(3,0)=-0.0883632752317; x(3,1)=-7.40468796501; x(3,2)=-3.07220055411;
+	x(4,0)=5.18231755738; x(4,1)=4.95152973815; x(4,2)=-9.9870338276;
+	x(5,0)=4.26708114291; x(5,1)=7.10165654603; x(5,2)=4.9253650409;
+	x(6,0)=8.18854988953; x(6,1)=2.12501081402; x(6,2)=0.879753850301;
+	x(7,0)=5.5827306913; x(7,1)=8.77507993975; x(7,2)=9.94739194247;
+	x(8,0)=6.51074058756; x(8,1)=9.82537500991; x(8,2)=-8.74331697256;
+	x(9,0)=-7.26254338011; x(9,1)=-9.73370985632; x(9,2)=8.32055931886;
+
+	CRegressionExample* aa=new CRegressionExample();
+
+	aa->set_x(x);
+	aa->set_y(y);
+	aa->set_init_w(w);
+	CMyStochasticCostFunction *fun=new CMyStochasticCostFunction();
+	fun->set_target(aa);
+
+	CFirstOrderPenalizedStochasticCostFunction* fun2=new CFirstOrderPenalizedStochasticCostFunction(fun);
+	fun2->set_penalty_weight(1.0);
+	CL2Penalty* penalty_type=new CL2Penalty();
+	fun2->set_penalty_type(penalty_type);
+
+	CSGDMinimizer* opt=new CSGDMinimizer(fun2);
+
+	CConstLearningRate* rate=new CConstLearningRate();
+	rate->set_const_learning_rate(0.0001);
+	CGradientDescendWithMomentumUpdater* updater=new CGradientDescendWithMomentumUpdater(rate);
+	updater->set_momentum(0.9);
+	opt->set_gradient_updater(updater);
+
+	int32_t num_passes=20;
+	opt->set_number_passes(num_passes);
+
+	opt->minimization();
+	//the result is from the svrg software using plain stochastic gradient descent
+	//http://riejohnson.com/svrg_download.html
+	//w=
+	//0.289383
+	//-1.513633
+	//2.045611
+	EXPECT_NEAR(w[0],0.289383,1e-6);
+	EXPECT_NEAR(w[1],-1.513633,1e-6);
+	EXPECT_NEAR(w[2],2.045611,1e-6);
+
+	SG_UNREF(opt);
+}
+
+TEST(SGDMinimizer,test4)
+{
+	SGVector<float64_t> w(3);
+	SGVector<float64_t> y(10);
+	SGMatrix<float64_t> x(10,3);
+	w.set_const(0.0);
+	y[0]=17.826341;
+	y[1]=28.947688;
+	y[2]=32.482436;
+	y[3]=5.475718;
+	y[4]=-26.082733;
+	y[5]=0.645608;
+	y[6]=1.794406;
+	y[7]=9.251004;
+	y[8]=-31.176166;
+	y[9]=30.801085;
+
+	x(0,0)=3.18934210549; x(0,1)=-6.36734839959; x(0,2)=3.87646568343;
+	x(1,0)=9.71965286623; x(1,1)=-6.49372199537; x(1,2)=7.85930917808;
+	x(2,0)=0.590611116182; x(2,1)=-9.78182994856; x(2,2)=8.12538297323;
+	x(3,0)=-0.0883632752317; x(3,1)=-7.40468796501; x(3,2)=-3.07220055411;
+	x(4,0)=5.18231755738; x(4,1)=4.95152973815; x(4,2)=-9.9870338276;
+	x(5,0)=4.26708114291; x(5,1)=7.10165654603; x(5,2)=4.9253650409;
+	x(6,0)=8.18854988953; x(6,1)=2.12501081402; x(6,2)=0.879753850301;
+	x(7,0)=5.5827306913; x(7,1)=8.77507993975; x(7,2)=9.94739194247;
+	x(8,0)=6.51074058756; x(8,1)=9.82537500991; x(8,2)=-8.74331697256;
+	x(9,0)=-7.26254338011; x(9,1)=-9.73370985632; x(9,2)=8.32055931886;
+
+	CRegressionExample* aa=new CRegressionExample();
+
+	aa->set_x(x);
+	aa->set_y(y);
+	aa->set_init_w(w);
+	CMyStochasticCostFunction *fun=new CMyStochasticCostFunction();
```

Comment:
step1 define/implement one stochastic cost function (that is a finite sum of sample-specific cost function)
eg:
f(w)=0.5* \sum_i{ (y_i-w'*x_i)^2 }, where (x_i,y_i) are features and a label for the i-th sample, w is the objective variable.